### PR TITLE
Allow MCSystemInterface::ListFolderEntries() of arbitrary paths.

### PR DESCRIPTION
**Pull request by peter-b**
Sometimes it's useful to iterate over the contents of directories
other than the current working directory.

N.b. I haven't been able to build & test all of these changes, so merging this might be risky. :neutral_face:

### CODE REVIEW

File: engine/src/mbliphone.mm [hunk 1]
```diff
@@ -815,10 +815,16 @@ virtual int64_t GetFileSize(void)
...
 {
+	MCStringRef t_path;
+	if (p_folder == nil)
+	  t_path = MCSTR(".");
+	else
+	  t_path = p_folder;
+
 	DIR *t_dir;
-	t_dir = opendir(".");
+	t_dir = opendir(MCStringGetCString(t_path));
```
Comment (runrevmark) [inline engine/src/mbliphone.mm]: This needs to MCAutoStringRefAsUTF8String'd

File: engine/src/mblandroidfs.cpp [hunk 2]
```diff
@@ -500,14 +496,21 @@ bool MCAndroidSystem::ResolvePath(MCStringRef p_path, MCStringRef& r_resolved)
...
+
+	MCStringRef t_path;
+	if (p_folder == nil)
+		t_path = MCSTR (".");
+	else
+		t_path = p_folder;
 
 	DIR *t_dir;
-	t_dir = opendir(".");
+	t_dir = opendir(MCStringGetCString (t_path));
```
Comment (runrevmark) [inline engine/src/mblandroidfs.cpp]: I think this also needs AsUTF8String'd

File: engine/src/dsklnx.cpp [hunk 3]
```diff
@@ -1912,9 +1912,15 @@ class MCLinuxDesktop: public MCSystemInterface
         closedir(dirptr);
         *dptr = tptr;
 #endif /* MCS_getentries_dsk_lnx */
+		MCStringRef t_path;
```
Comment (runrevmark) [inline engine/src/dsklnx.cpp]: This needs to be MCAutoStringRefAsSysString'd :)

**Comment by peter-b** [comment]
This patch series as it stands doesn't work because most of the platforms' implementations subsequently call:

```
struct dirent *t_dir_entry;
struct stat t_stat;
t_dir_entry = readdir(t_dir);
stat(t_dir_entry->d_name, &t_stat);
```

Obviously, the call to `stat` won't work correctly if the directory of `t_dir` isn't the current working directory.

File: engine/src/mblandroidfs.cpp [hunk 4]
```diff
@@ -115,16 +115,12 @@ bool apk_set_current_folder(MCStringRef p_apk_path)
...
 }
 
-bool apk_list_folder_entries(MCSystemListFolderEntriesCallback p_callback, void *p_context)
+bool apk_list_folder_entries(MCStringRef p_apk_folder, MCSystemListFolderEntriesCallback p_callback, void *p_context)
 {
 	bool t_success = true;
 	MCAutoStringRef t_list;
-    MCAutoStringRef t_current_folder;
-    
-    if (!apk_get_current_folder(&t_current_folder))
```
Comment (livecodefraser) [inline engine/src/mblandroidfs.cpp]: This call is still needed when the APK folder isn't passed in

**Comment by livecodefraser** [comment]
@livecode-vulcan reviewed okay 1c5fb2c

**Comment by livecode-vulcan** [comment]
💙 review by @livecodefraser ok 1c5fb2c04fa469e68804d5f0d551aece3792fd42

**Comment by livecode-vulcan** [comment]
😞 test failure 1c5fb2c04fa469e68804d5f0d551aece3792fd42
- try-linux-x86: [failure](https://vulcan.livecode.com/builders/try-linux-x86/builds/203)
- try-win-x86: [failure](https://vulcan.livecode.com/builders/try-win-x86/builds/267)
- try-ios: unknown\* try-android-armv6: [failure](https://vulcan.livecode.com/builders/try-android-armv6/builds/214)
- try-linux-x86_64: [failure](https://vulcan.livecode.com/builders/try-linux-x86_64/builds/149)
- try-mac: unknown

**Comment by peter-b** [comment]
@livecode-vulcan try 1c5fb2c

**Comment by livecode-vulcan** [comment]
💥 merge conflict 1c5fb2c04fa469e68804d5f0d551aece3792fd42 → 3e4576d5d0699a30680ae4bbb230ac6a5833dfbf

**Comment by peter-b** [comment]
@livecode-vulcan reviewed by @runrevmark ok 596ad32

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok 596ad3239b0ecce1a89e51d41e779978f8b9fada

**Comment by runrevmark** [comment]
@livecode-vulcan review ok f3780bd

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok f3780bd61a26c71ade5282b16267d410efc4f2aa

**Comment by livecode-vulcan** [comment]
😞 test failure f3780bd61a26c71ade5282b16267d410efc4f2aa
- try-linux-x86: [success](https://vulcan.livecode.com/builders/try-linux-x86/builds/452)
- try-win-x86: [pending](https://vulcan.livecode.com/builders/try-win-x86/builds/512)
- try-ios: [success](https://vulcan.livecode.com/builders/try-ios/builds/398)
- try-android-armv6: [failure](https://vulcan.livecode.com/builders/try-android-armv6/builds/463)
- try-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-linux-x86_64/builds/398)
- try-mac: [success](https://vulcan.livecode.com/builders/try-mac/builds/389)

**Comment by livecodeali** [comment]
@livecode-vulcan review ok c5bd220

**Comment by livecode-vulcan** [comment]
💙 review by @livecodeali ok c5bd220aa40f7b515717b97b21f3ef1468009fb4

**Comment by livecode-vulcan** [comment]
😎 test success c5bd220aa40f7b515717b97b21f3ef1468009fb4
- try-linux-x86: [success](https://vulcan.livecode.com/builders/try-linux-x86/builds/641)
- try-win-x86: [success](https://vulcan.livecode.com/builders/try-win-x86/builds/701)
- try-ios: [success](https://vulcan.livecode.com/builders/try-ios/builds/590)
- try-android-armv6: [success](https://vulcan.livecode.com/builders/try-android-armv6/builds/654)
- try-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-linux-x86_64/builds/586)
- try-mac: [success](https://vulcan.livecode.com/builders/try-mac/builds/579)

**Pull request merged**